pomdp_py/framework/basics.pyx

```diff
@@ -317,33 +411,132 @@ cdef class Observation:
     def __ne__(self, other):
         return not self.__eq__(other)
 
+cdef class Vector(list):
+    """
+    The Vector class. Provides an implementation of a vector for multi-valued response models.
+    """
+    def __init__(self, values=list()):
+        if not isinstance(values, list):
+            raise TypeError(f"values must be type list, but got {type(values)}.")
+        for v in values:
+            self.append(v)
+
+    def __eq__(self, other):
+        if not isinstance(other, (Vector, list)):
+            raise TypeError(f"other must be type Vector or list, but got {type(other)}.")
+        return len(self) == len(other) and all(v0 == v1 for v0, v1 in zip(self, other))
+
+    def __add__(self, other):
+        if isinstance(other, (float, int)):
+            vec = [other] * len(self)
+        elif isinstance(other, Vector):
+            vec = other
+        else:
+            raise TypeError(f"other must be type Vector, float, or int, but got {type(other)}.")
+        return Vector([v0 + v1 for v0, v1 in zip(self, vec)])
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __mul__(self, other):
+        if not isinstance(other, (float, int)):
+            raise TypeError(f"other must be type float or int, but got {type(other)}.")
+        return Vector([v * other for v in self])
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+
+cdef class Response:
+    """
+    A Response class that only handles a scalar reward. Subclasses of Response can add
+    more (scalar or vector) variables. But the subclasses must implement how to handle
+    arithmetic and comparison operations.
```

Comment:
I see. The idea of vector reward is interesting, and I value the contribution. But currently the changes are not backwards compatible. I suggest the following changes:

* Create a file `pomdp_py/framework/generalization.pyx` in which:
    *  Define a class called `ResponseAgent` that inherits `Agent` but it takes in a response model instead of a reward model.
    * Define `Response`, `ResponseModel` and other necessary components there.
    * Define a `sample_generative_function` specific for `ResponseAgent`. That way you don't need to make changes to the existing code using float-based reward.
* Create a file `pomdp_py/algorithms/cc_pomcp.pyx` in which you can implement the CC-POMCP algorithm.

Reply:
Hey @zkytony, thanks for the notes. Your suggestion is helpful.

Ideally, I would have liked to reuse methods from your PO-UCT and POMCP implementations in the CC-POMCP algorithm. However, I cannot do so via subclassing because PO-UCT and POMCP return a single value in `rollout` and `simulate`, while CC-POMCP returns two values.

To avoid a backwards compatibility issue, I could reproduce most of the PO-UCT and POMCP code in CC-POMCP. Please let me know how this approach sounds.

Reply:
Can you subclass and override rollout and simulate?

Reply:
@troiwill does that make sense? You can still let CC-POMCP inherit PO-UCT, but override `rollout` and `simulate` so they can internally work with multi-valued rewards. You might want to create custom `VNode` and/or `QNode` classes, since the value stored in the nodes are also not just floats, I suppose.

Reply:
@troiwill does that make sense? You can still let CC-POMCP inherit PO-UCT, but override rollout and simulate so they can internally work with multi-valued rewards. You might want to create custom VNode and/or QNode classes, since the value stored in the nodes are also not just floats, I suppose.

Reply:
Hey @zkytony, that makes sense. Once I've made all the changes, I will create another pull request. Thanks for the note.